Isolate `sasl` as a dependency

Right now, `pip install dbt-spark` is not working on Windows because:
- `dbt-spark` requires `PyHive[hive]`
- `hive` extra requires `sasl`
- `pip install sasl` is failing on Windows machines

On my personal machine:
```
Python 3.8.5
pip 20.2.3
```

Here's one of the stacktraces I've seen, though they vary by user/machine:
```
Building wheels for collected packages: sasl
  Building wheel for sasl (setup.py) ... error
  ERROR: Command errored out with exit status 1:
...
Complete output (27 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.7
  creating build\lib.win-amd64-3.7\sasl
  copying sasl\__init__.py -> build\lib.win-amd64-3.7\sasl
  running egg_info
  writing sasl.egg-info\PKG-INFO
  writing dependency_links to sasl.egg-info\dependency_links.txt
  writing requirements to sasl.egg-info\requires.txt
  writing top-level names to sasl.egg-info\top_level.txt
  reading manifest file 'sasl.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'sasl.egg-info\SOURCES.txt'
  copying sasl\saslwrapper.cpp -> build\lib.win-amd64-3.7\sasl
  copying sasl\saslwrapper.h -> build\lib.win-amd64-3.7\sasl
  copying sasl\saslwrapper.pyx -> build\lib.win-amd64-3.7\sasl
  running build_ext
  building 'sasl.saslwrapper' extension
  creating build\temp.win-amd64-3.7
  creating build\temp.win-amd64-3.7\Release
  creating build\temp.win-amd64-3.7\Release\sasl
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Isasl -Ic:\users\User\appdata\local\programs\python\python37\include -Ic:\users\User\appdata\local\programs\python\python37\include "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" "-IC:\Program Files (x86)\Windows Kits\10\include\wdf\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\wdf\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\wdf\um" "-IC:\Program Files (x86)\Windows Kits\10\include\wdf\winrt" /EHsc /Tpsasl/saslwrapper.cpp /Fobuild\temp.win-amd64-3.7\Release\sasl/saslwrapper.obj 
  saslwrapper.cpp
  c:\users\User\appdata\local\programs\python\python37\include\pyconfig.h(59): fatal error C1083: Cannot open include file: 'io.h': No such file or directory   
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\x86_amd64\\cl.exe' failed with exit status 2
----------------------------------------
ERROR: Failed building wheel for sasl
  Running setup.py clean for sasl
Failed to build sasl
Installing collected packages: pure-sasl, thrift-sasl, sasl, PyHive, dbt-spark
    Running setup.py install for sasl ... error
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,8 @@
 ## dbt-spark 0.19.0 (Release TBD)
 
+### Breaking changes
+- Users of the `http` and `thrift` connection methods need to install extra requirements: `pip install dbt-spark[PyHive]` ([#109](https://github.com/fishtown-analytics/dbt-spark/pull/109), [#126](https://github.com/fishtown-analytics/dbt-spark/pull/126))
+
 ### Under the hood
 - Add changelog, issue templates ([#119](https://github.com/fishtown-analytics/dbt-spark/pull/119), [#120](https://github.com/fishtown-analytics/dbt-spark/pull/120))
 

--- a/dbt/adapters/spark/connections.py
+++ b/dbt/adapters/spark/connections.py
@@ -8,9 +8,14 @@ from dbt.logger import GLOBAL_LOGGER as logger
 from dbt.utils import DECIMALS
 from dbt.adapters.spark import __version__
 
-from TCLIService.ttypes import TOperationState as ThriftState
-from thrift.transport import THttpClient
-from pyhive import hive
+try:
+    from TCLIService.ttypes import TOperationState as ThriftState
+    from thrift.transport import THttpClient
+    from pyhive import hive
+except ImportError:
+    ThriftState = None
+    THttpClient = None
+    hive = None
 try:
     import pyodbc
 except ImportError:
@@ -86,6 +91,19 @@ class SparkCredentials(Credentials):
             raise dbt.exceptions.RuntimeException(
                 "`cluster` and `endpoint` cannot both be set when"
                 f" using {self.method} method to connect to Spark"
+            )
+
+        if (
+            self.method == SparkConnectionMethod.HTTP or
+            self.method == SparkConnectionMethod.THRIFT
+        ) and not (
+            ThriftState and THttpClient and hive
+        ):
+            raise dbt.exceptions.RuntimeException(
+                f"{self.method} connection method requires "
+                "additional dependencies. \n"
+                "Install the additional required dependencies with "
+                "`pip install dbt-spark[PyHive]`"
             )
 
     @property

--- a/setup.py
+++ b/setup.py
@@ -61,11 +61,13 @@ setup(
     },
     install_requires=[
         f'dbt-core=={dbt_version}',
-        'PyHive[hive]>=0.6.0,<0.7.0',
         'sqlparams>=3.0.0',
-        'thrift>=0.11.0,<0.12.0'
     ],
     extras_require={
         "ODBC":  ['pyodbc>=4.0.30'],
+        "PyHive":  [
+            'PyHive[hive]>=0.6.0,<0.7.0',
+            'thrift>=0.11.0,<0.12.0',
+        ],
     }
 )

